lib/src/collecting_body.rs

```diff
@@ -0,0 +1,340 @@
+    chunks: Vec<Bytes>,
+    error: Option<Error>,
+    trailers: Option<HeaderMap>,
+}
+
+impl CollectingBody {
+    /// Create a new CollectingBody that stores & streams from the provided Body.
+    ///
+    /// Writes to the StreamingBody are collected, and propagated to all readers of this
+    /// CollectingBody.
+    // TODO: Expected length?
+    pub fn new(from: Body) -> CollectingBody {
+        let (tx, rx) = watch::channel(CollectedBody::default());
+        let body = CollectingBody { inner: rx };
+        tokio::task::spawn(Self::tee(from, tx));
+        body
+    }
+
+    /// "tee" a single Body to the watch channel.
+    ///
+    /// This is the worker thread behind a CollectingBody. It reads the data from the provided body
+    /// into a `tokio::sync::watch` channel, which (a) accumulates the body + trailers and (b)
+    /// notifies any subscribed readers of the updates. The readers can safely miss updates or
+    /// start late, as they always can eventually read the state.
+    async fn tee(mut rx: Body, tx: watch::Sender<CollectedBody>) {
+        // IMPORTANT IMPLEMENTATION NOTE:
+        //
+        // Make sure every path out of this function results in either state.errors.is_some() or
+        // state.trailers.is_some().
+
+        // Read data first:
+        while let Some(chunk) = rx.data().await {
+            match chunk {
+                Ok(data) => {
+                    tx.send_modify(move |state| {
+                        state.chunks.push(data);
+                    });
+                }
+                Err(Error::Again) => continue,
+                Err(e) => {
+                    tx.send_modify(move |state| {
+                        state.error = Some(e);
+                    });
+                    return;
+                }
+            }
+        }
+
+        // Then wait for trailers (if any) to be present:
+        let trailers = rx.trailers().await;
+        tx.send_modify(move |state| match trailers {
+            Ok(v) => state.trailers = Some(v.unwrap_or_default()),
+            Err(e) => state.error = Some(e),
+        });
+    }
+
+    /// Get a new read handle to this body.
+    pub fn read(&self) -> Result<Body, error::Error> {
+        let mut upstream = self.inner.clone();
+        let (mut tx, rx) = StreamingBody::new();
+        tokio::task::spawn(async move {
+            let mut next_chunk = 0;
+
+            loop {
+                // We can't hold current_value across an await point,
+                // and borrowck doesn't undersand drop() enough for us to use it in a branch.
+                // Queue up our work, then do it.
+                let send_chunks: Vec<Bytes>;
+                let trailers: Option<HeaderMap>;
+                let error: Option<String>;
+                if let Ok(current_value) = upstream
+                    .wait_for(|v| {
+                        v.trailers.is_some() || v.error.is_some() || v.chunks.len() > next_chunk
```

Comment:
This can always be true, so you can switch to `.changed().await`, if you only call `tx.send_modify()` to push new data when the chunk of data has at least one byte up on L80.

Reply:
… I guess it already is always true, but you could collect zero-size chunks which only advance `next_chunk`